src/pages/workspaces/hooks/useDeletionPolling.ts

```diff
@@ -0,0 +1,120 @@
+import _ from 'lodash/fp';
+import { useEffect, useState } from 'react';
+import { Ajax } from 'src/libs/ajax';
+import { workspacesStore, workspaceStore } from 'src/libs/state';
+import { pollWithCancellation } from 'src/libs/utils';
+import { WorkspaceState, WorkspaceWrapper as Workspace } from 'src/libs/workspace-utils';
+
+const updateWorkspacesList = (
+  workspaces: Workspace[],
+  workspaceId: string,
+  state: WorkspaceState,
+  errorMessage?: string
+): Workspace[] => workspaces.map((ws) => updateWorkspace(ws, workspaceId, state, errorMessage));
+
+const updateWorkspace = (
+  workspace: Workspace,
+  workspaceId: string,
+  state: WorkspaceState,
+  errorMessage?: string
+): Workspace => {
+  if (workspace.workspace.workspaceId === workspaceId) {
+    const update = _.cloneDeep(workspace);
+    update.workspace.state = state;
+    update.workspace.errorMessage = errorMessage;
+    return update;
+  }
+  return workspace;
+};
+
+const doUpdate = (abort: () => void, workspace: Workspace, state: WorkspaceState, errorMessage?: string) => {
+  const workspaceId = workspace.workspace.workspaceId;
+  abort();
+  workspacesStore.update((wsList) => updateWorkspacesList(wsList, workspaceId, state, errorMessage));
+  workspaceStore.update((ws) => updateWorkspace(ws, workspaceId, state, errorMessage));
+};
+
+export const useDeletionPolling = (workspaces: Workspace[]) => {
+  // we have to do the signal/abort manually instead of with useCancelable so that the it can be cleaned up in the
+  // this component's useEffect, instead of the useEffect in useCancelable
+  const [controller, setController] = useState(new window.AbortController());
+  const abort = () => {
+    controller.abort();
+    setController(new window.AbortController());
+  };
+
+  useEffect(() => {
+    const checkWorkspaceDeletion = async (workspace: Workspace) => {
```

Comment:
`checkWorkspaceDeletion` can be shared (that's what I intended with my previous comment).